### PR TITLE
PagedAttention Transformation: Rank alignment for replacements

### PR DESCRIPTION
**Pull request by slyalin**
During the elimination of dependencies from `beam_idx` input and `ReadValue`(s), we are replacing them by the new PA-related inputs and sub-expressions dependent on other remaining inputs. In such replacements we need to guarantee matching shape and element type of old and new nodes. Before this PR it was not guaranteed for shape and sometimes a scalar was replaced by a shape of rank 1 that led to errors like `'start' input is not a scalar`. Now the shape is aligned.

### CODE REVIEW

**Review by ilya-lavrenov — APPROVED** [review]
Confirm, that expected models are fixed by this PR.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/prev_sequence_length_pattern.cpp [hunk 1]
```diff
@@ -39,30 +38,23 @@ ov::pass::PrevSequenceLengthPattern::PrevSequenceLengthPattern(
...
-            }
-            replace_node(gather, replacement);
-            return true;
+            replacement = prev_max_seq_len;
+        } else {
+            // assumption that any other axis should point to batch dimension, precise reasoning is too complex
+            // TODO: provide more reliable check
+            replacement = batch_dim;
         }
+        if (replacement->get_output_element_type(0) != target_type) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/prev_sequence_length_pattern.cpp:49]: is it possible that we will have f16 -> fp32, int -> fp conversions here?
Could this cause any issues in accuracy?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/prev_sequence_length_pattern.cpp [hunk 2]
```diff
@@ -39,30 +38,23 @@ ov::pass::PrevSequenceLengthPattern::PrevSequenceLengthPattern(
...
+            replacement = std::make_shared<v0::Convert>(replacement, target_type);
+        }
+        auto required_shape = gather->get_output_partial_shape(0);
+        if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/prev_sequence_length_pattern.cpp:53]: minor: It's better to change order of the checks:

required_shape.rank().is_static() && replacement->get_output_partial_shape(0) != required_shape

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 3]
```diff
@@ -29,11 +30,13 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
...
-        if (max_context_len->get_output_element_type(0) != target_type) {
-            replacement = std::make_shared<v0::Convert>(max_context_len, target_type);
-        } else {
-            replacement = max_context_len;
+        std::shared_ptr<Node> replacement = max_context_len;
+        if (replacement->get_output_element_type(0) != target_type) {
+            replacement = std::make_shared<v0::Convert>(replacement, target_type);
+        }
+        auto required_shape = gather->get_output_partial_shape(0);
+        if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:38]: minor: It's better to change order of the checks:

required_shape.rank().is_static() && replacement->get_output_partial_shape(0) != required_shape

**Comment by itikhono** [comment]
@CuriousPanCake did we test these changes via https://github.com/openvinotoolkit/openvino/pull/24634?

**Comment by CuriousPanCake** [comment]
> @CuriousPanCake did we test these changes via #24634?

Not yet. I'll test ASAP

**Comment by ilya-lavrenov** [comment]
> @CuriousPanCake did we test these changes via #24634?

See "Confirm, that expected models are fixed by this PR." above

**Review by itikhono — APPROVED** [review]

**Review by CuriousPanCake — APPROVED** [review]

**Pull request merged**